### PR TITLE
change basefee to be a module param

### PR DESCRIPTION
**Pull request by thomas-nguy**
Closes: #937

## Description

Change the logic to make base fee a module param in order to be able to change it through governance proposal.

This is an alternative for https://github.com/tharsis/ethermint/pull/938 

Pro : Easier to maintain
Con : Breaking change

Might also have slight degradation in term of performance compared to storing in state storage but we don't need to use event to get the base fee at height anymore 

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: proto/ethermint/feemarket/v1/feemarket.proto [hunk 1]
```diff
@@ -13,8 +13,8 @@ message Params {
   // elasticity multiplier bounds the maximum gas limit an EIP-1559 block may
   // have.
   uint32 elasticity_multiplier = 3;
-  // initial base fee for EIP-1559 blocks.
-  int64 initial_base_fee = 4;
+  // base fee for EIP-1559 blocks.
+  string base_fee = 4;
```
Comment (fedekunze) [inline proto/ethermint/feemarket/v1/feemarket.proto:17]: use customtype to convert to sdk.Int

File: x/feemarket/types/params.go [hunk 2]
```diff
@@ -106,14 +113,19 @@ func validateElasticityMultiplier(i interface{}) error {
 	return nil
 }
 
-func validateInitialBaseFee(i interface{}) error {
-	value, ok := i.(int64)
+func validateBaseFee(i interface{}) error {
+	value, ok := i.(string)
```
Comment (fedekunze) [inline x/feemarket/types/params.go:117]: this should be sdk.Int

File: x/feemarket/keeper/params.go [hunk 3]
```diff
@@ -16,3 +17,27 @@ func (k Keeper) GetParams(ctx sdk.Context) (params types.Params) {
...
+	if !valid {
+		return nil
+	}
+
+	return baseFee
+}
+
+// SetBaseFee set's the base fee in the paramSpace
+func (k Keeper) SetBaseFee(ctx sdk.Context, baseFee *big.Int) {
+	k.paramSpace.Set(ctx, types.ParamStoreKeyBaseFee, baseFee.String())
```
Comment (fedekunze) [inline x/feemarket/keeper/params.go:42]: why another paramstore?

File: x/feemarket/keeper/grpc_query.go [hunk 4]
```diff
@@ -21,10 +21,18 @@ func (k Keeper) Params(c context.Context, _ *types.QueryParamsRequest) (*types.Q
 }
 
 // BaseFee implements the Query/BaseFee gRPC method
+// return empty if base fee is not enable
```
Comment (fedekunze) [inline x/feemarket/keeper/grpc_query.go:24]:
```suggestion
// return empty if base fee is not enabled
```

File: x/feemarket/keeper/eip1559.go [hunk 5]
```diff
@@ -25,13 +25,13 @@ func (k Keeper) CalculateBaseFee(ctx sdk.Context) *big.Int {
...
 	if ctx.BlockHeight() == params.EnableHeight {
-		return new(big.Int).SetInt64(params.InitialBaseFee)
+		return k.GetBaseFee(ctx)
 	}
 
 	// get the block gas used and the base fee values for the parent block.
 	parentBaseFee := k.GetBaseFee(ctx)
 	if parentBaseFee == nil {
-		parentBaseFee = new(big.Int).SetInt64(params.InitialBaseFee)
+		return nil
```
Comment (fedekunze) [inline x/feemarket/keeper/eip1559.go:34]: why this change?

File: x/feemarket/keeper/eip1559.go [hunk 6]
```diff
@@ -43,7 +43,7 @@ func (k Keeper) CalculateBaseFee(ctx sdk.Context) *big.Int {
 
 	parentGasTargetBig := new(big.Int).Div(gasLimit, new(big.Int).SetUint64(uint64(params.ElasticityMultiplier)))
 	if !parentGasTargetBig.IsUint64() {
-		return new(big.Int).SetInt64(params.InitialBaseFee)
+		return nil
```
Comment (fedekunze) [inline x/feemarket/keeper/eip1559.go:46]: why this change?

File: x/feemarket/keeper/abci.go [hunk 7]
```diff
@@ -19,14 +17,6 @@ func (k *Keeper) BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) {
...
 
 	k.SetBaseFee(ctx, baseFee)
-
-	// Store current base fee in event
-	ctx.EventManager().EmitEvents(sdk.Events{
-		sdk.NewEvent(
-			types.EventTypeFeeMarket,
-			sdk.NewAttribute(types.AttributeKeyBaseFee, baseFee.String()),
-		),
-	})
```
Comment (fedekunze) [inline x/feemarket/keeper/abci.go:29]: I think we should still emit this event for clients that would like to track the base fee

File: proto/ethermint/feemarket/v1/genesis.proto [hunk 8]
```diff
@@ -10,9 +10,9 @@ option go_package = "github.com/tharsis/ethermint/x/feemarket/types";
 message GenesisState {
   // params defines all the paramaters of the module.
   Params params = 1 [ (gogoproto.nullable) = false ];
-  // base fee is the exported value from previous software version.
+  // default base fee is the exported value from previous software version.
   // Zero by default.
-  string base_fee = 2 [
+  string default_base_fee = 2 [
```
Comment (fedekunze) [inline proto/ethermint/feemarket/v1/genesis.proto:15]: we should remove this field as it will be included in the params

**Review by fedekunze — CHANGES_REQUESTED** [review]

File: x/feemarket/keeper/abci.go (hunk 7 above)
Comment (thomas-nguy) [reply x/feemarket/keeper/abci.go:29]: yes I also realized we need to keep retrocompatible, I will add it back

File: x/feemarket/keeper/eip1559.go (hunk 5 above)
Comment (thomas-nguy) [reply x/feemarket/keeper/eip1559.go:34]: does it make sense to return `params.InitialBaseFee` here?  

if parentbasefee cannot be found, should the function calculate return nil and disable basefee?

File: x/feemarket/keeper/eip1559.go (hunk 6 above)
Comment (thomas-nguy) [reply x/feemarket/keeper/eip1559.go:46]: same here, does it make sense to use `params.InitialBaseFee` when the parameter may be wrongly set?

File: x/feemarket/keeper/eip1559.go (hunk 5 above)
Comment (fedekunze) [reply x/feemarket/keeper/eip1559.go:34]: the default behavior on geth is to return the initial base fee afaik, if the base fee is not found then it we can assume it should have been checked before?

File: proto/ethermint/feemarket/v1/feemarket.proto [hunk 9]
```diff
@@ -13,8 +15,11 @@ message Params {
   // elasticity multiplier bounds the maximum gas limit an EIP-1559 block may
   // have.
   uint32 elasticity_multiplier = 3;
-  // initial base fee for EIP-1559 blocks.
-  int64 initial_base_fee = 4;
+  // base fee for EIP-1559 blocks.
+  string base_fee = 4 [
```
Comment (fedekunze) [inline proto/ethermint/feemarket/v1/feemarket.proto:19]: you need to use `6` here and use `reserved` for the `initial_base_fee`.

https://github.com/tharsis/ethermint/blob/bd0f073704336700f6b3e8cdabaf2842216ab7ee/proto/ethermint/evm/v1/evm.proto#L102-L104

Comment (fedekunze) [reply proto/ethermint/feemarket/v1/feemarket.proto:19]: cc: @danburck @yihuang @loredanacirstea @ramacarlucho @crypto-facs 

This is relevant in case you are modifying an existing field.

File: proto/ethermint/feemarket/v1/genesis.proto [hunk 10]
```diff
@@ -10,13 +10,7 @@ option go_package = "github.com/tharsis/ethermint/x/feemarket/types";
...
-  // base fee is the exported value from previous software version.
-  // Zero by default.
-  string base_fee = 2 [
-    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
-    (gogoproto.nullable) = false
-  ];
   // block gas is the amount of gas used on the last block before the upgrade.
   // Zero by default.
-  uint64 block_gas = 3;
+  uint64 block_gas = 2;
```
Comment (fedekunze) [inline proto/ethermint/feemarket/v1/genesis.proto:15]: same, use `reserved` for `base_fee` and keep the `3` for `block_gas`

File: x/feemarket/keeper/params.go [hunk 11]
```diff
@@ -16,3 +17,24 @@ func (k Keeper) GetParams(ctx sdk.Context) (params types.Params) {
...
+		return nil
+	}
+
+	return params.BaseFee.BigInt()
+}
+
+// SetBaseFee set's the base fee in the paramSpace
+func (k Keeper) SetBaseFee(ctx sdk.Context, baseFee *big.Int) {
+	k.paramSpace.Set(ctx, types.ParamStoreKeyBaseFee, sdk.NewIntFromBigInt(baseFee))
+}
```
Comment (fedekunze) [inline x/feemarket/keeper/params.go:40]: remove and use `SetParams` instead

**Review by fedekunze — CHANGES_REQUESTED** [review]

File: x/feemarket/keeper/eip1559.go (hunk 5 above)
Comment (loredanacirstea) [reply x/feemarket/keeper/eip1559.go:34]: go-ethereum checks this in the `VerifyEip1559Header ` function
https://github.com/ethereum/go-ethereum/blob/53b1420edefcf6be154ea6df887b1e4d68fcc36f/consensus/misc/eip1559.go#L41-L44, as part of the `verifyHeader` function - e.g. in https://github.com/ethereum/go-ethereum/blob/b3b8b268eb585dfd3c1c9e9bbebc55968f3bec4b/consensus/ethash/consensus.go#L300-L303

Are block headers checked by ethermint or geth?

File: proto/ethermint/feemarket/v1/feemarket.proto (hunk 9 above)
Comment (thomas-nguy) [reply proto/ethermint/feemarket/v1/feemarket.proto]: it should be set accordingly to your network traffic and pricing strategy. 
`elasticity_multiplier`  defines how much the gas price can vary between blocks and `base_fee_change_denominator` defines your target gas used per block.

File: x/feemarket/keeper/params.go (hunk 11 above)
Comment (thomas-nguy) [reply x/feemarket/keeper/params.go:40]: This is a shortcut for 

```
params := k.getParams(ctx)
params.BaseFee = sdk.NewIntFromBigInt(baseFee)
SetParams(params)
```

Is there any issue using it?

Comment (fedekunze) [reply x/feemarket/keeper/params.go:40]: no, I guess it's fine

File: x/feemarket/keeper/eip1559.go (hunk 5 above)
Comment (thomas-nguy) [reply x/feemarket/keeper/eip1559.go:34]: afaik, we don't validate `header.basefee` on london block on ethermint.

Geth uses a strong assumption that parent base fee cannot be nil thanks to the validation, thus it is not checked during the calculation
https://github.com/ethereum/go-ethereum/blob/53b1420edefcf6be154ea6df887b1e4d68fcc36f/consensus/misc/eip1559.go#L68

 But in our case, we have different config for enabling feemarket and setting london blocks number which are not necessarily corollated by the code itself.
 
 I guess we need to make a decision in that case. 
 
 I can return the "param.baseFee" here if it is the behavior that we want to have, let me know what you think

File: x/feemarket/keeper/eip1559.go [hunk 12]
```diff
@@ -25,13 +25,13 @@ func (k Keeper) CalculateBaseFee(ctx sdk.Context) *big.Int {
 
 	// If the current block is the first EIP-1559 block, return the InitialBaseFee.
 	if ctx.BlockHeight() == params.EnableHeight {
-		return new(big.Int).SetInt64(params.InitialBaseFee)
+		return k.GetBaseFee(ctx)
 	}
 
 	// get the block gas used and the base fee values for the parent block.
 	parentBaseFee := k.GetBaseFee(ctx)
```
Comment (fedekunze) [inline x/feemarket/keeper/eip1559.go:32]: no need to get the params from the store again, it's already available and we just need to convert it to `big.Int`

File: x/feemarket/keeper/eip1559.go (hunk 5 above)
Comment (fedekunze) [reply x/feemarket/keeper/eip1559.go:34]: we can return the default param base fee value for consistency with geth

Comment (thomas-nguy) [reply x/feemarket/keeper/eip1559.go:34]: actually since we store the base fee as a module param, this line does not make more sense anymore

since `k.GetBaseFee(ctx)` and `paramsBaseFee` return the same things.

do you means we should 
- either use 0 value for parentBaseFee
- introduce a new parameter in genesis default_base_fee and use it here?

**Review by fedekunze — APPROVED** [review]

**Comment by fedekunze** [comment]
rpc tests are failing

File: tests/rpc/rpc_test.go [hunk 13]
```diff
@@ -565,7 +565,7 @@ func TestEth_EthResend(t *testing.T) {
 func TestEth_FeeHistory(t *testing.T) {
 	params := make([]interface{}, 0)
 	params = append(params, 4)
-	params = append(params, "0x1c")
+	params = append(params, "0xa")
```
Comment (thomas-nguy) [inline tests/rpc/rpc_test.go:568]: slightly modify to a lower value in case the chain is running behind or the integration tests are running too fast

**Review by crypto-facs — APPROVED** [review]
LGTM

File: x/feemarket/keeper/params.go [hunk 14]
```diff
@@ -16,3 +17,24 @@ func (k Keeper) GetParams(ctx sdk.Context) (params types.Params) {
 func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
 	k.paramSpace.SetParamSet(ctx, &params)
 }
+
+// ----------------------------------------------------------------------------
+// Parent Base Fee
+// Required by EIP1559 base fee calculation.
+// ----------------------------------------------------------------------------
+
+// GetConstantFee get's the base fee from the paramSpace
```
Comment (yihuang) [inline x/feemarket/keeper/params.go:26]:
```suggestion
// GetBaseFee gets the base fee from the paramSpace
```

File: x/feemarket/keeper/params.go [hunk 15]
```diff
@@ -16,3 +17,24 @@ func (k Keeper) GetParams(ctx sdk.Context) (params types.Params) {
...
+// return nil if base fee is not enabled
+func (k Keeper) GetBaseFee(ctx sdk.Context) *big.Int {
+	params := k.GetParams(ctx)
+	if params.NoBaseFee {
+		return nil
+	}
```
Comment (yihuang) [inline x/feemarket/keeper/params.go:32]: Should we also check `EnableHeight` here?

Comment (thomas-nguy) [reply x/feemarket/keeper/params.go:32]: hm not sure 🤔 

Those two parameters seems redundant. 

My understanding is `EnableHeight` controls the calculation while `NoBaseFee` controls the basefee as a whole

**Comment by fedekunze** [comment]
@thomas-nguy @yihuang Can you help out with the migration PR? 🙏

**Comment by thomas-nguy** [comment]
yeah sure but I think the migration is not that important since the json-rpc endpoint still relies on the event to get base fee value which did not change from previous software version

**Comment by yihuang** [comment]
> yeah sure but I think the migration is not that important since the json-rpc endpoint still relies on the event to get base fee value which did not change from previous software version

maybe the only migration needed is to delete the base fee key from the state?
